File: android/src/main/kotlin/com/example/nearby_cross/callbacks/NearbyCrossCallbacks.kt
package com.example.nearby_cross.callbacks

import com.example.nearby_cross.constants.ChannelMethods
import io.flutter.plugin.common.MethodChannel

class NearbyCrossCallbacks(private val channel: MethodChannel) {
    val advertiser = object : AdvertiserCallbacks() {
        override fun onPayloadReceived(bytesReceived: ByteArray, endpointId: String) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["message"] = bytesReceived
            hashmap["endpointId"] = endpointId.toByteArray()
            channel.invokeMethod(ChannelMethods.PAYLOAD_RECEIVED, hashmap)
        }

        override fun onConnectionInitiated(
            endpointId: String,
            endpointName: ByteArray,
            alreadyAcceptedConnection: Boolean
        ) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            hashmap["endpointName"] = endpointName
            hashmap["alreadyAcceptedConnection"] =
                if (alreadyAcceptedConnection) "1".toByteArray() else "0".toByteArray()
            channel.invokeMethod(ChannelMethods.CONNECTION_INITIATED, hashmap)
        }

        override fun onSuccessfulConnection(endpointId: String) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            channel.invokeMethod(ChannelMethods.SUCCESSFUL_CONNECTION, hashmap)
        }

        override fun onRejectedConnection(endpointId: String) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            channel.invokeMethod(ChannelMethods.CONNECTION_REJECTED, hashmap)
        }

        override fun onDisconnected(endpointId: String) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            channel.invokeMethod(ChannelMethods.ENDPOINT_DISCONNECTED, hashmap)
        }
    }

    val discoverer = object : DiscovererCallbacks() {
        override fun onEndpointFound(endpointId: String, endpointName: ByteArray) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            hashmap["endpointName"] = endpointName
            channel.invokeMethod(ChannelMethods.ON_ENDPOINT_FOUND, hashmap)
        }

        override fun onEndpointLost(endpointId: String) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            channel.invokeMethod(ChannelMethods.ON_ENDPOINT_LOST, hashmap)
        }

        override fun onPayloadReceived(bytesReceived: ByteArray, endpointId: String) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["message"] = bytesReceived
            hashmap["endpointId"] = endpointId.toByteArray()
            channel.invokeMethod(ChannelMethods.PAYLOAD_RECEIVED, hashmap)
        }

        override fun onConnectionInitiated(
            endpointId: String,
            endpointName: ByteArray,
            alreadyAcceptedConnection: Boolean
        ) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            hashmap["endpointName"] = endpointName
            hashmap["alreadyAcceptedConnection"] =
                if (alreadyAcceptedConnection) "1".toByteArray() else "0".toByteArray()
            channel.invokeMethod(ChannelMethods.CONNECTION_INITIATED, hashmap)
        }

        override fun onSuccessfulConnection(endpointId: String) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            channel.invokeMethod(ChannelMethods.SUCCESSFUL_CONNECTION, hashmap)
        }

        override fun onRejectedConnection(endpointId: String) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            channel.invokeMethod(ChannelMethods.CONNECTION_REJECTED, hashmap)
        }

        override fun onDisconnected(endpointId: String) {
            val hashmap = HashMap<String, ByteArray>()
            hashmap["endpointId"] = endpointId.toByteArray()
            channel.invokeMethod(ChannelMethods.ENDPOINT_DISCONNECTED, hashmap)
        }
    }
}

File: ios/Classes/callbacks/NearbyCrossCallbacks.swift
//
//  NearbyCrossCallbacks.swift
//  nearby_cross
//
//  Created by Ignacio Velasco on 29/02/2024.
//

import Foundation


import Flutter

class NearbyCrossCallbacks {
    let channel: FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    lazy var advertiser: AdvertiserCallbacks = {
        return AdvertiserCallbacksImpl(channel: channel)
    }()
    
    lazy var discoverer: DiscovererCallbacks = {
        return DiscovererCallbacksImpl(channel: channel)
    }()
}

class AdvertiserCallbacksImpl: AdvertiserCallbacks {
    let channel: FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    func onPayloadReceived(bytesReceived: Data, endpointId: String) {
        var hashmap = [String: Data]()
        hashmap["message"] = bytesReceived
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        channel.invokeMethod(ChannelMethods.PAYLOAD_RECEIVED, arguments: hashmap)
    }

    func onConnectionInitiated(endpointId: String, endpointName: Data, alreadyAcceptedConnection: Bool) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        hashmap["endpointName"] = endpointName
        hashmap["alreadyAcceptedConnection"] = alreadyAcceptedConnection ? "1".data(using: .utf8) : "0".data(using: .utf8)
        channel.invokeMethod(ChannelMethods.CONNECTION_INITIATED, arguments: hashmap)
    }

    func onSuccessfulConnection(endpointId: String) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        channel.invokeMethod(ChannelMethods.SUCCESSFUL_CONNECTION, arguments: hashmap)
    }
    
    func onRejectedConnection(endpointId: String) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        channel.invokeMethod(ChannelMethods.CONNECTION_REJECTED, arguments: hashmap)
    }

    func onDisconnected(endpointId: String) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        channel.invokeMethod(ChannelMethods.ENDPOINT_DISCONNECTED, arguments: hashmap)
    }
}

class DiscovererCallbacksImpl: DiscovererCallbacks {
    let channel: FlutterMethodChannel
    
    init(channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    func onPayloadReceived(bytesReceived: Data, endpointId: String) {
        var hashmap = [String: Data]()
        hashmap["message"] = bytesReceived
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        channel.invokeMethod(ChannelMethods.PAYLOAD_RECEIVED, arguments: hashmap)
    }

    func onConnectionInitiated(endpointId: String, endpointName: Data, alreadyAcceptedConnection: Bool) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        hashmap["endpointName"] = endpointName
        hashmap["alreadyAcceptedConnection"] = alreadyAcceptedConnection ? "1".data(using: .utf8) : "0".data(using: .utf8)
        channel.invokeMethod(ChannelMethods.CONNECTION_INITIATED, arguments: hashmap)
    }

    func onSuccessfulConnection(endpointId: String) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        channel.invokeMethod(ChannelMethods.SUCCESSFUL_CONNECTION, arguments: hashmap)
    }
    
    func onRejectedConnection(endpointId: String) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        channel.invokeMethod(ChannelMethods.CONNECTION_REJECTED, arguments: hashmap)
    }

    func onDisconnected(endpointId: String) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        channel.invokeMethod(ChannelMethods.ENDPOINT_DISCONNECTED, arguments: hashmap)
    }
    
    func onEndpointFound(endpointId: String, endpointName: Data) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        hashmap["endpointName"] = endpointName
        channel.invokeMethod(ChannelMethods.ON_ENDPOINT_FOUND, arguments: hashmap)
    }

    func onEndpointLost(endpointId: String) {
        var hashmap = [String: Data]()
        hashmap["endpointId"] = endpointId.data(using: .utf8)
        channel.invokeMethod(ChannelMethods.ON_ENDPOINT_LOST, arguments: hashmap)
    }
}
